tensorly/contrib/decomposition/tt_TTOI.py

```diff
@@ -0,0 +1,188 @@
+    Reference paper: Zhou Y, Zhang AR, Zheng L, Wang Y. "Optimal high-order tensor svd via tensor-train orthogonal iteration."
+
+    Parameters
+    ----------
+    data_tensor: tl.tensor
+        observed tensor data
+    rank : tuple
+        rank of the TT decomposition
+        must verify rank[0] == rank[-1] == 1 (boundary conditions)
+        and len(rank) == len(tl.shape(data_tensor))+1
+    n_iter : int
+        half the number of iterations
+    trajectory : bool, optional, default is False
+        if True, the output of each iteration of TTOI is returned: 2*n_iter outputs
+        otherwise, only the output of the last iteration is returned
+    return_errors : bool, optional, default is True
+        if True, the approximation/reconstruction error of each iteration of TTOI is returned: 2*n_iter outputs
+    Returns
+    -------
+    factors : list of n_iter tensors or one tensor
+    * n_iter tensors (if `trajectory` is True) : each list contains the output of each iteration, one full tensor and list of tensor factors
+    * one tensor (otherwise): output of the last iteration, one full tensor and list of tensor factors 
+    full_tensor : list of n_iter tensors or one tensor
+    * n_iter tensors (if `trajectory` is True) : each list contains the output of each iteration, one full tensor and list of tensor factoros
+    * one tensor (otherwise): output of the last iteration, one full tensor and list of tensor factors 
+    """
+    context = tl.context(data_tensor)
+    shape = tl.shape(data_tensor)
+    n_dim = len(shape) 
+
+    rank = validate_tt_rank(shape, rank)
+
+    # Make sure it's not a tuple but a list
+    rank = list(rank)
+    
+    # Add two one-dimensional mode to data_tensor
+    data_tensor_extended = tl.reshape(data_tensor,(1, ) + shape + (1, ))
+
+
+    if trajectory:
+        factors = list()
+        full_tensor = list()
+        
+    if return_errors:
+        error_list = list()
+        
+    # perform TTOI for n_iter iterations
+    for n in range(n_iter):
+        # first perform forward update
+        # U_arr will be a list including estimated left singular spaces at the current iteration
+        U_arr = list()
+        # initialize R_tilde_arr (sequential unfolding of data_tensor multiplied by U_arr sequentially on the left, useful for backward update to obtain V_arr)
+        R_tilde_arr = list()
+
+        # estimate the first left singular spaces
+        # Here, R_tmp is the first sequential unfolding compressed on the right by previous updated V_arr (if exists)
+        R_tmp_l = data_tensor_extended
+        if n == 0:
+            R_tmp = R_tmp_l
+        else:
+            R_tmp = sequential_prod(R_tmp_l,V_arr,"right")
+        U_tmp = tl.partial_svd(tl.reshape(R_tmp,(shape[0],-1)),rank[1])[0]
+        U_arr.append(tl.reshape(U_tmp,(rank[0],shape[0],rank[1])))
+
+        # estimate the 2nd to (d-1)th left singular spaces
+        for k in range(n_dim-2):
```

Comment:
```suggestion
        for mode in range(n_dim-2):
```
same here, good to have explicit names